### PR TITLE
nixos/bitcoind: change to multi-instance + add tests

### PR DESCRIPTION
**Pull request by 1000101**
###### Motivation for this change
Currently, it is not possible to run multiple instances of bitcoind (typical use case is running mainnet and testnet on the same server). While at it, I also refactored some things which I think weren't properly configured (e.g. RPC options were ignored when running testnet) or deprecated (loaOf). Plus I have also added basic tests.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 1000101** [comment]
@GrahamcOfBorg test bitcoind

**Comment by 1000101** [comment]
@GrahamcOfBorg test bitcoind

**Comment by mmahut** [comment]
Please add this change to the release notes for the next release.

File: nixos/modules/services/networking/bitcoind.nix [hunk 1]
```diff
@@ -149,45 +140,118 @@ in {
...
     };
-    users.users.${cfg.user} = {
+  };
+
+  config = mkIf (eachBitcoind != {}) {
+
+    assertions = flatten (mapAttrsToList (bitcoindName: cfg: [
+    {
+      assertion = (cfg.prune != null) -> (cfg.prune == "disable" || cfg.prune == "manual") || (cfg.prune == 0 || cfg.prune == 1) || (cfg.prune >= 550);
+      message = "If set, services.bitcoind.${bitcoindName}.prune has to be "disable", "manual", 0 , 1 or >= 550.";
```
Comment (prusnak) [inline nixos/modules/services/networking/bitcoind.nix:173]:
```suggestion
      message = "If set, services.bitcoind.${bitcoindName}.prune has to be \"disable\", \"manual\", 0 , 1 or >= 550.";
```

File: nixos/modules/services/networking/bitcoind.nix [hunk 2]
```diff
@@ -149,45 +140,118 @@ in {
...
+      description = "Specification of one or more bitcoind instances.";
     };
-    users.users.${cfg.user} = {
+  };
+
+  config = mkIf (eachBitcoind != {}) {
+
+    assertions = flatten (mapAttrsToList (bitcoindName: cfg: [
+    {
+      assertion = (cfg.prune != null) -> (cfg.prune == "disable" || cfg.prune == "manual") || (cfg.prune == 0 || cfg.prune == 1) || (cfg.prune >= 550);
```
Comment (prusnak) [inline nixos/modules/services/networking/bitcoind.nix:172]:
```suggestion
      assertion = (cfg.prune != null) -> (cfg.prune == "disable" || cfg.prune == "manual" || cfg.prune == 0 || cfg.prune == 1 || cfg.prune >= 550);
```

**Review by prusnak — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/bitcoind.nix (hunk 2 above)
Comment (1000101) [reply nixos/modules/services/networking/bitcoind.nix:172]: Nice.

File: nixos/modules/services/networking/bitcoind.nix (hunk 1 above)
Comment (1000101) [reply nixos/modules/services/networking/bitcoind.nix:173]: Ah, thanks, I introduced the error during refactoring (making multi-line comment a single-line).

**Comment by 1000101** [comment]
@GrahamcOfBorg test bitcoind

**Comment by 1000101** [comment]
@GrahamcOfBorg test bitcoind

**Comment by aanderse** [comment]
This dramatically increases the complexity of this module. Who is the maintainer? Are containers not an adequate way to run multiple instances of this service on the same machine?

@mmahut @prusnak is the added complexity "worth it"?

**Comment by mmahut** [comment]
@aanderse hmm, good point, but I don't consider this to be a major rewrite, at least running both testnet and mainnet networks at the same time is a standard feature that has been omitted in the previous module. I like the rewrite, the previous code was a mess (understandable, if we consider what a mess the upstream bitcoin node is).

@1000101 as mentioned, please note this change in the release notes and if possible, add yourself as maintainer (I guess you are fine with it given you added yourself to the test?).

**Comment by aanderse** [comment]
@mmahut thanks for feedback.

**Comment by 1000101** [comment]
> This dramatically increases the complexity of this module. Who is the maintainer? Are containers not an adequate way to run multiple instances of this service on the same machine?
> 
> @mmahut @prusnak is the added complexity "worth it"?

It may look like it, but it's just a rewrite. In fact, everything will work out of box (as previously), all you have to do is name the instance and that's it. I've updated some of stuff I found not working properly but other than that, I kept it compatible with the original service (apart from the instance name).

**Comment by 1000101** [comment]
> @aanderse hmm, good point, but I don't consider this to be a major rewrite, at least running both testnet and mainnet networks at the same time is a standard feature that has been omitted in the previous module. I like the rewrite, the previous code was a mess (understandable, if we consider what a mess the upstream bitcoin node is).
> 
> @1000101 as mentioned, please note this change in the release notes and if possible, add yourself as maintainer (I guess you are fine with it given you added yourself to the test?).

@mmahut yeap, will do

**Comment by 1000101** [comment]
@GrahamcOfBorg test bitcoind

**Comment by mmahut** [comment]
Looks good, thank you.

**Comment by emmanuelrosa** [comment]
This changes the bitcoind data directory, correct? That means that if there's already a running instance which is using `/var/lib/bitcoind/` as the data directory, this change would ignore that and bitcoind would start doing an initial sync.

There's also the issue of changing the bitcoind user, since existing files would be owned by the original bitcoind user.

You may need to use NixOS `stateVersion` (and accompaning release notes) to prevent this change from applying automatically to users which already run bitcoind.

**Comment by mmahut** [comment]
Given this is breaking the way this service is configured I think just mentioning it in the release notes is enough, given this change will not redefine the directory nor user out of the box without users intervention.

But you have a good point that it would be useful if the release notes did mention this change, just in case. @1000101 do you mind opening an addendum?

**Comment by emmanuelrosa** [comment]
To use this new multi-instance config with an existing bitcoind data directory, the user would have to change from something like this:

```
  services.bitcoind = {
    enable = true;
    extraConfig = "...";
    ...
  };
```

To something like this:

```
  services.bitcoind.mainnet = {
    enable = true;
    dataDir = "/var/lib/bitcoind";
    user = "bitcoin";
    extraConfig = "...";
    ...
  };
```
The key settings are:

 *  `dataDir` - to continue using the same data directory
 *  `user` - to continue using the same user so that bitcoind maintains access to its files.

**Comment by 1000101** [comment]
> This changes the bitcoind data directory, correct? That means that if there's already a running instance which is using `/var/lib/bitcoind/` as the data directory, this change would ignore that and bitcoind would start doing an initial sync.
> 
> There's also the issue of changing the bitcoind user, since existing files would be owned by the original bitcoind user.
> 
The directory and user may/will change by default, but you have the option to set them manually to the original (desired) values. However, you are absolutely right and I should mention this in the release notes (incompatibility section).

> You may need to use NixOS `stateVersion` (and accompaning release notes) to prevent this change from applying automatically to users which already run bitcoind.

Since it is a breaking change it will not automatically change, however, I definitely need to add the points you mention in the release notes.

**Comment by 1000101** [comment]
> To use this new multi-instance config with an existing bitcoind data directory, the user would have to change from something like this:
> 
> ```
>   services.bitcoind = {
>     enable = true;
>     extraConfig = "...";
>     ...
>   };
> ```
> 
> To something like this:
> 
> ```
>   services.bitcoind.mainnet = {
>     enable = true;
>     dataDir = "/var/lib/bitcoind";
>     user = "bitcoin";
>     extraConfig = "...";
>     ...
>   };
> ```
> 
> The key settings are:
> 
>     * `dataDir` - to continue using the same data directory
> 
>     * `user` - to continue using the same user so that bitcoind maintains access to its files.

Thank you so much, I'll amend it.

File: nixos/doc/manual/release-notes/rl-2009.xml [hunk 3]
```diff
@@ -531,6 +531,23 @@ systemd.services.nginx.serviceConfig.ReadWritePaths = [ "/var/www" ];
      to be used for every display-manager in NixOS.
     </para>
    </listitem>
+  <listitem>
```
Comment (aanderse) [inline nixos/doc/manual/release-notes/rl-2009.xml:534]: Should we add migration steps for existing users?

File: nixos/modules/services/networking/bitcoind.nix [hunk 4]
```diff
@@ -110,29 +94,36 @@ in {
...
-          type = with types; loaOf (submodule rpcUserOpts);
-          description = ''
-            RPC user information for JSON-RPC connnections.
-          '';
+          type = types.attrsOf (types.submodule rpcUserOpts);
+          description = "RPC user information for JSON-RPC connnections.";
         };
       };
 
+      pidFile = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/networking/bitcoind.nix:102]: Why does this need to be configurable?

**Review by aanderse — COMMENTED** [review]
Out of curiosity would it make sense to use `DynamicUser` at all here?

File: nixos/doc/manual/release-notes/rl-2009.xml (hunk 3 above)
Comment (1000101) [reply nixos/doc/manual/release-notes/rl-2009.xml:534]: Absolutely, have you seen https://github.com/NixOS/nixpkgs/pull/94166 ? - I've scratched the original notes completely, as this section should document the incompatibility, so it's only relevant to people already using the service.

File: nixos/modules/services/networking/bitcoind.nix (hunk 4 above)
Comment (1000101) [reply nixos/modules/services/networking/bitcoind.nix:102]: It kind of [was there before](https://github.com/NixOS/nixpkgs/blame/55079328c219fb53f3120a34fdba3a80d3be601b/nixos/modules/services/networking/bitcoind.nix#L29), so I didn't want to scratch it completely. I was aiming for as least incompatibility as I could. But you're right and it doesn't need to be that way.

File: nixos/doc/manual/release-notes/rl-2009.xml (hunk 3 above)
Comment (aanderse) [reply nixos/doc/manual/release-notes/rl-2009.xml:534]: I hadn't seen. Thanks!

**Comment by emmanuelrosa** [comment]
If `DynamicUser` changes the user ID then I don't recommend using it (for users already using bitcoind) because then `bitcoind` may not be able to access the files in it's data directory.

**Comment by aanderse** [comment]
`DynamicUser` keeps all data under `StateDirectory` owned by the user the service runs by, even when that user changes.

**Pull request merged**